File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/SignDataViewController.m
//
//  SignDataViewController.m
//  
//
//  Created by Federal Authority For Identity and Citizenship  on 1/23/17.
//  Copyright © 2017 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "SignDataViewController.h"

@interface SignDataViewController ()  {
    uint8_t *resultSignature;
    int resultSignatureLength;
}
@property(nonatomic,strong)Utils * utils;
@end

@implementation SignDataViewController
#pragma mark viewDidLoad

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.enterPinText.keyboardType=UIKeyboardTypeNumberPad;
    self.enterPinText.keyboardAppearance = UIKeyboardAppearanceLight;
    UIToolbar* enterPinNumberToolbar = [[UIToolbar alloc]initWithFrame:CGRectMake(0, 0, 320, 50)];
    enterPinNumberToolbar.barStyle = UIBarStyleBlackTranslucent;
    enterPinNumberToolbar.items = @[[[UIBarButtonItem alloc]initWithTitle:@"Cancel" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinCancelNumberPad)],
                                    [[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil],
                                    [[UIBarButtonItem alloc]initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinDoneWithNumberPad)]];
    [enterPinNumberToolbar sizeToFit];
    _enterPinText.inputAccessoryView = enterPinNumberToolbar;
    
    self.title=@"Sign Data";
    self.navigationItem.hidesBackButton = YES;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(goBack)];
    
    self.signDataButton.layer.cornerRadius=22.0;
    [self.signDataButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.signDataButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.signDataButton.layer setShadowOpacity:0.5];
 //   self.signDataButton.layer.borderWidth=1.0;
    
    self.verifySignatureButton.layer.cornerRadius=22.0;
    [self.verifySignatureButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.verifySignatureButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.verifySignatureButton.layer setShadowOpacity:0.5];
  //  self.verifySignatureButton.layer.borderWidth=1.0;
    
    SwitchValue =1;
    
    AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    self.model = aDelegate.model;
    self.utils=aDelegate.utils;
    
    // Do any additional setup after loading the view.
}

#pragma mark viewWillAppear

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:YES];
    
    self.enterPinText.text=@"";
    self.digitalSignTextView.text=@"";
    self.enterDataTextView.text=@"";
    [self.enterPinText resignFirstResponder];
    [self.enterDataTextView resignFirstResponder];

    self.digitalSignTextView.text=@"";
    
    self.enterPinText.enabled=YES;
    self.enterDataTextView.editable=YES;
    self.enterDataTextView.selectable=YES;
}
#pragma mark EIDA Toolkit calling functions

- (IBAction)signDataButtonAction:(id)sender {
    
    self.digitalSignTextView.text=@"";
    [self.enterPinText resignFirstResponder];
    [self.enterDataTextView resignFirstResponder];
    
    if ([self.enterPinText.text length]==0) {
        [AlertView showAlertTitle:ALERT withMessage:ENTER_PIN onView:self];
    }// if
    else {
        if ([self.enterPinText.text length]<4) {
            [AlertView showAlertTitle:ALERT withMessage:MIN_PIN onView:self];
            self.enterPinText.text=@"";
        }// if
        else if ([self.enterPinText.text length]>16) {
            [AlertView showAlertTitle:ALERT withMessage:MAX_PIN onView:self];
            self.enterPinText.text=@"";
        }// else if
        else {
        [self.utils ShowProgressBar:SIGNDATA andView:self.view];
        int data_hash = 0;  /// 0 for plain data and 1 for Hash data
        NSString *requestId=[Utils generateSecureKey];
        NSString *textviewData =self.enterDataTextView.text;
        dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
        @try {
            CardReader *cardreader  = [self.model.cardReaderSharedArray objectAtIndex:0];
            DataProtectionKey *data_protection_key=self.model.dataprotectedShared;
            
            NSString *request_handle =  [cardreader prepareRequest:requestId];
            if (request_handle.length==0 || [request_handle isEqualToString:@""] || [request_handle isEqualToString:@"(null)"] ) {
                [AlertView showAlertTitle:ALERT withMessage:REQUEST_HANDLE_EMPTY onView:self];
            }//if
            else {
                NSString *encodePin=[Utils setEncrytion:request_handle data:self.enterPinText.text publickey:[data_protection_key getPublicKey] keylength:[data_protection_key getKeyLength]];
                if (encodePin.length==0 || [encodePin isEqualToString:@""] || [encodePin isEqualToString:@"(null)"] ) {
                    [AlertView showAlertTitle:ALERT withMessage:ENCODE_PIN_EMPTY onView:self];
                }//if
                else {
                    SignatureResponse *signresponse;
                    if (self->SwitchValue ==1) {
                        signresponse = [cardreader signChallenge:textviewData inputLength:(int)[textviewData length] isInputHash:data_hash encodedPin:encodePin];
                    }//if
                    else if (self->SwitchValue==2) {
                        signresponse = [cardreader signData:self.enterDataTextView.text inputLength:(int)[self.enterDataTextView.text length] isInputHash:data_hash encodedPin:encodePin];
                    }//else if
                dispatch_async(dispatch_get_main_queue(), ^{
                    NSString *xmlString =[signresponse getXmlString];
                    NSString *errorMsg = [self.utils validateToolkitResponse:requestId xmlstring:xmlString];
                    if (![errorMsg isEqualToString:@""] || errorMsg.length>0) {
                        [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                    }//if
                    else {
                        self->resultSignature=[signresponse getSignature];
                        self->resultSignatureLength =[signresponse getSignatureLength];
                        NSData *csignData = [[NSData alloc] initWithBytes:self->resultSignature length:self->resultSignatureLength];
                        NSString *signDataBase64string=[Utils base64forData:csignData];
    
                        self.digitalSignTextView.text=signDataBase64string;
                     }//else
                   });// update UI main queue
                 }//else
               }//else
            }//try
            @catch (NSException *exception) {
                NSString *vgresponse =exception.userInfo[@"ErrorResponse"];
                if (![vgresponse isEqualToString:@""] && vgresponse.length>0) {
                    NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:vgresponse];
                    if (![errorMsg isEqualToString:@""] && errorMsg.length>0) {
                        [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                    }//if
                    else {
                        NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                        [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
                    }//else
                }//if
                else {
                    NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                    [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
                }
            }// catch
            @finally {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.utils DismissProgressBar];
                });// update UI main queue
            }// finally
         });// background queue
     }//else
  }//else
}
- (IBAction)verifySignatureButtonAction:(id)sender {
    
   [self.utils ShowProgressBar:VERIFYDATA andView:self.view];
   [self.enterPinText resignFirstResponder];
   [self.enterDataTextView resignFirstResponder];
    
    NSString *requestId=[Utils generateSecureKey];
    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
    @try {
        CardReader *cardreader  = [self.model.cardReaderSharedArray objectAtIndex:0];
        DataProtectionKey *data_protection_key=self.model.dataprotectedShared;
        
        NSString *request_handle =  [cardreader prepareRequest:requestId];
        if (request_handle.length==0 || [request_handle isEqualToString:@""] || [request_handle isEqualToString:@"(null)"] ) {
            [AlertView showAlertTitle:ALERT withMessage:REQUEST_HANDLE_EMPTY onView:self];
        }//if
        else {
            NSString *encodePin=[Utils setEncrytion:request_handle data:self.enterPinText.text publickey:[data_protection_key getPublicKey] keylength:[data_protection_key getKeyLength]];
            if (encodePin.length==0 || [encodePin isEqualToString:@""] || [encodePin isEqualToString:@"(null)"] ) {
                [AlertView showAlertTitle:ALERT withMessage:ENCODE_PIN_EMPTY onView:self];
            }//if
            else {
                CardCertificates *certificates = [cardreader getPkiCertificates:encodePin];
                NSString *xmlString =[certificates getXmlString];
                NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:xmlString];
                if (![errorMsg isEqualToString:@""] || errorMsg.length>0) {
                    [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                }//if
                else {
                    int data_hash = 0;  /// 0 for plain data and 1 for Hash data
                    uint8_t *certificate;
                    int certificateLength;
                        
                    if (self->SwitchValue ==1) {
                        certificate = [certificates getAuthenticationCertificate];
                        certificateLength= [certificates getAuthenticationCertificateLength];
                    }//if
                    else if (self->SwitchValue==2) {
                        certificate = [certificates getSigningCertificate];
                        certificateLength= [certificates getSigningCertificateLength];
                    }//else if
                    [cardreader verifySignature:self.enterDataTextView.text inputLength:(int)[self.enterDataTextView.text length] isInputHash:data_hash signature:self->resultSignature signatureLength:self->resultSignatureLength certificate:certificate certificateLength:certificateLength];
                    [AlertView showAlertTitle:ALERT withMessage:SIGNATURE_VERIFIED onView:self ];
                }//else
            }//else
        }//else
    }//try
        @catch (NSException *exception) {
            NSString *vgresponse =exception.userInfo[@"ErrorResponse"];
            if (![vgresponse isEqualToString:@""] && vgresponse.length>0) {
                NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:vgresponse];
                if (![errorMsg isEqualToString:@""] && errorMsg.length>0) {
                    [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                }//if
                else {
                    NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                    [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
                }//else
            }//if
            else {
                NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
            }
        }// catch
    @finally {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.utils DismissProgressBar];
        });// update UI main queue
      }// finally
   });// background queue
}
- (IBAction)swithButton:(id)sender {
    if ([self.swithButton isOn]) {
        self.authKey.text = @"Sign Key";
        [self.swithButton setOn:NO animated:YES];
        SwitchValue=2;
        [self.signDataButton setTitle:@"Sign With Sign Key" forState:UIControlStateNormal];
        [self.verifySignatureButton setTitle:@"Verify With Sign Key" forState:UIControlStateNormal];
        self.enterPinText.text=@"";
        self.enterDataTextView.text=@"";
        self.digitalSignTextView.text=@"";
    }//if
    else {
        self.authKey.text = @"Auth Key";
        [self.swithButton setOn:YES animated:YES];
        SwitchValue=1;
        [self.signDataButton setTitle:@"Sign With Auth Key" forState:UIControlStateNormal];
        [self.verifySignatureButton setTitle:@"Verify With Auth Key" forState:UIControlStateNormal];
        self.enterPinText.text=@"";
        self.enterDataTextView.text=@"";
        self.digitalSignTextView.text=@"";
    }//else
}
#pragma mark Updating UI Methods
- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
-(void)enterPinCancelNumberPad {
    [_enterPinText resignFirstResponder];
    _enterPinText.text = @"";
}
-(void)enterPinDoneWithNumberPad {
    [_enterPinText resignFirstResponder];
}
- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text {
    
    if ([text isEqualToString:@"\n"]) {
        
        [textView resignFirstResponder];
        // Return FALSE so that the final '\n' character doesn't get added
        return NO;
    }
    // For any other character return TRUE so that the text gets added to the view
    return YES;
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/SignDataViewController.swift
//
//  SignDataViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 09/05/22.
//

import UIKit

class SignDataViewController: UIViewController {
    
    @IBOutlet var enterDataTextView: UITextView!
    @IBOutlet weak var enterPinText: UITextField!
    @IBOutlet var digitalSignTextView: UITextView!
    @IBOutlet weak var authKey: UILabel!
    @IBOutlet var signDataButton: UIButton!
    @IBOutlet var verifySignatureButton: UIButton!
    @IBOutlet weak var view_Roundedcorners: UIView!
    @IBOutlet var switchButton: UISwitch!
    var  SwitchValue:Int = 0

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)
        
        self.enterPinText.keyboardType = .numberPad
        self.enterPinText.keyboardAppearance = .light
        let enterPinNumberToolbar = UIToolbar(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(320), height: CGFloat(50)))
        enterPinNumberToolbar.barStyle = .blackTranslucent
        enterPinNumberToolbar.items = [UIBarButtonItem(title: "Cancel", style: .done, target: self, action: #selector(enterPinCancelNumberPad)), UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil), UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(enterPinDoneWithNumberPad))]
        enterPinNumberToolbar.sizeToFit()
        self.enterPinText.inputAccessoryView=enterPinNumberToolbar

        // Do any additional setup after loading the view.
        
        signDataButton.layer.cornerRadius = 22.0
        signDataButton.layer.shadowOffset = CGSize(width: 5.0, height: 5.0)
        signDataButton.layer.shadowColor = UIColor.lightGray.cgColor
        signDataButton.layer.shadowOpacity = 0.5
        
        verifySignatureButton.layer.cornerRadius = 22.0
        verifySignatureButton.layer.shadowOffset = CGSize(width: 5.0, height: 5.0)
        verifySignatureButton.layer.shadowColor = UIColor.lightGray.cgColor
        verifySignatureButton.layer.shadowOpacity = 0.5
        
        SwitchValue = 1
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        
        enterPinText.text = ""
        enterDataTextView.text = ""
        enterPinText.resignFirstResponder()
        enterDataTextView.resignFirstResponder()
        digitalSignTextView.text = ""
        enterPinText.isEnabled = true
        enterDataTextView.isEditable = true
        enterDataTextView.isEditable = true
        
    }
    
    @IBAction func signDataButtonAction(_ sender: Any) {
                
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        else if cardreader == nil {
            self.view.makeToast(Constants.cardreaderNotConnect)
            return
        }//else if
        
        digitalSignTextView.text = ""
        self.enterPinText.resignFirstResponder()
        self.enterDataTextView.resignFirstResponder()
        if self.enterPinText.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.EnterPin, onView: self)
        }// if
        else {
            if (self.enterPinText.text?.count)!<4 {
                AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MinPin, onView: self)
                self.enterPinText.text=""
            }// if
            else if (self.enterPinText.text?.count)!>16 {
                AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MaxPin, onView: self)
                self.enterPinText.text=""
            }// else if
            else {
                utils.showProgressBar(Constants.SignData, andView: self.view)
                let requestId = Utils.generateSecureKey()
                let dataHash: Bool = false  /// 0 for plain data and 1 for Hash data
                let pinText = self.enterPinText.text!
                let enteredDataText = self.enterDataTextView.text!
                DispatchQueue.global(qos: .background).async {
                    do {
                        let request_handle = try cardreader?.prepareRequest(requestId!)
                        if (request_handle?.isEmpty == true) || (request_handle == "") || (request_handle == "(null)") {
                            AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Request_Handle_Empty, onView: self)
                        }//if
                        else {
                            let encodePin:String? = Utils.setEncrytion(request_handle, data: pinText,publickey: dataProtectionKey.getPublicKey(), keylength: Int32(dataProtectionKey.getKeyLength()))
                            if (encodePin?.isEmpty)! || (encodePin == "") || (encodePin == "(null)") {
                                AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Encode_Pin_Empty, onView: self)
                            }//if
                            else {
                                if self.SwitchValue == 1 {
                                    signresponse = try cardreader?.signChallenge(enteredDataText, inputLength: UInt32(enteredDataText.count), isInputHash: dataHash, encodedPin: encodePin!)
                                }//if
                                else if (self.SwitchValue == 2) {
                                    signresponse = try cardreader?.signData(enteredDataText, inputLength: UInt32(enteredDataText.count), isInputHash: dataHash, encodedPin: encodePin!)
                                }//else if
                                DispatchQueue.main.async {
                                    let xmlString = signresponse.getXmlString()
                                    let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: xmlString)
                                    if (errorMsg != "") && (errorMsg.count ) > 0 {
                                        AlertView.showAlertTitle(Constants.Alert, withMessage: errorMsg, onView: self)
                                    }//if
                                    else {
                                        let csignData = NSData(bytes: signresponse.getSignature(), length: Int(signresponse.getSignatureLength()))
                                        let signDataBase64string=Utils.base64forData(csignData as Data)
                                        self.digitalSignTextView.text=signDataBase64string
                                    }//else
                                }// update UI main queue
                            }//else
                        }//else
                        DispatchQueue.main.async {
                            utils.dismissProgressBar()
                        }// update UI main queue
                    }//do
                    catch let error as NSError{
                        DispatchQueue.main.async {
                            var vgresponse = String()
                            if error.userInfo["ErrorResponse"] != nil {
                                vgresponse = error.userInfo["ErrorResponse"] as! String
                                if (vgresponse != "") && ((vgresponse.count ) > 0) {
                                    let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: vgresponse)
                                    if (errorMsg != "") && ((errorMsg.count ) > 0) {
                                        AlertView.showAlertTitle(Constants.Alert, withMessage:errorMsg, onView: self)
                                    }//if
                                    else {
                                        let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                        AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                                    }//else
                                }//if
                                else {
                                    let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                    AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                                }//else
                            }//if
                            else {
                                let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                            }//else
                            utils.dismissProgressBar()
                            
                            self.enterDataTextView.text=""
                            self.enterPinText.text=""
                        }// update UI main queue
                    }//catch
                }//background queue
            }//else
        }//else
    }
    
    
    @IBAction func verifySignatureButtonAction(_ sender: Any) {
        
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        else if cardreader == nil {
            self.view.makeToast(Constants.cardreaderNotConnect)
            return
        }//else if
        
        utils.showProgressBar(Constants.VerifyData, andView: self.view)
        let requestId = Utils.generateSecureKey()
        let dataHash: Bool = false  /// 0 for plain data and 1 for Hash data
        let pinText = self.enterPinText.text!
        let dataText = self.enterDataTextView.text
        let dataTextCount = self.enterDataTextView.text.count
        DispatchQueue.global(qos: .background).async {
            do {
                let request_handle = try cardreader?.prepareRequest(requestId!)
                if (request_handle?.isEmpty == true) || (request_handle == "") || (request_handle == "(null)") {
                    AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Request_Handle_Empty, onView: self)
                }//if
                else {
                    let encodePin:String? = Utils.setEncrytion(request_handle, data: pinText,publickey: dataProtectionKey.getPublicKey(), keylength: Int32(dataProtectionKey.getKeyLength()))
                    if (encodePin?.isEmpty)! || (encodePin == "") || (encodePin == "(null)") {
                        AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Encode_Pin_Empty, onView: self)
                    }//if
                    else {
                        certificates = try cardreader?.getPkiCertificates(encodePin!)
                        let xmlString = certificates.getXmlString()
                        let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: xmlString)
                        if (errorMsg != "") && (errorMsg.count ) > 0 {
                            AlertView.showAlertTitle(Constants.Alert, withMessage: errorMsg, onView: self)
                        }//if
                        else {
                            if self.SwitchValue == 1 {
                                try cardreader?.verifySignature(dataText!, inputLength: UInt32(dataTextCount), isInputHash: dataHash, signature: signresponse.getSignature() , signatureLength: signresponse.getSignatureLength(), certificate: certificates.getAuthenticationCertificate(), certificateLength: certificates.getAuthenticationCertificateLength())
                            }//if
                            else if (self.SwitchValue == 2) {
                                try cardreader?.verifySignature(dataText!, inputLength: UInt32(dataTextCount), isInputHash: dataHash, signature: signresponse.getSignature() , signatureLength: signresponse.getSignatureLength(), certificate: certificates.getSigningCertificate(), certificateLength: certificates.getSigningCertificateLength())
                            }//else if
                            AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.SignatureVerified,onView: self)
                        }//else
                    }//else
                }//else
                DispatchQueue.main.async {
                    utils.dismissProgressBar()
                }// update UI main queue
            }//do
            catch let error as NSError{
                DispatchQueue.main.async {
                    var vgresponse = String()
                    if error.userInfo["ErrorResponse"] != nil {
                        vgresponse = error.userInfo["ErrorResponse"] as! String
                        if (vgresponse != "") && ((vgresponse.count ) > 0) {
                            let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: vgresponse)
                            if (errorMsg != "") && ((errorMsg.count ) > 0) {
                                AlertView.showAlertTitle(Constants.Alert, withMessage:errorMsg, onView: self)
                            }//if
                            else {
                                let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                            }//else
                        }//if
                        else {
                            let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                            AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                        }//else
                    }//if
                    else {
                        let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                        AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                    }//else
                    
                    utils.dismissProgressBar()
                }// update UI main queue
            }//catch
        }//background queue
    }
    
    @IBAction func switchButtonAction(_ sender: Any) {
        
        if SwitchValue == 1 {
            authKey.text = "Sign Key"
            switchButton.setOn(false, animated: true)
            SwitchValue = 2
            signDataButton.setTitle("Sign With Sign Key", for: .normal)
            verifySignatureButton.setTitle("Verify With Sign Key", for: .normal)
            enterPinText.text = ""
            enterDataTextView.text = ""
            digitalSignTextView.text = ""
        }
        else {
            authKey.text = "Auth Key"
            switchButton.setOn(true, animated: true)
            SwitchValue = 1
            signDataButton.setTitle("Sign With Auth Key", for: .normal)
            verifySignatureButton.setTitle("Verify With Auth Key", for: .normal)
            enterPinText.text = ""
            enterDataTextView.text = ""
            digitalSignTextView.text = ""
        }
    }

    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }

    @objc func enterPinCancelNumberPad() {
        enterPinText.resignFirstResponder()
        enterPinText.text=""
    }
    
    @objc func enterPinDoneWithNumberPad() {
        enterPinText.resignFirstResponder()
    }
    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        
        if text == "\n" {
            textView.resignFirstResponder()
            return false
        }
        return true
    }
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
